src/VirtoCommerce.CartModule.Web/BackgroundJobs/DeleteObsoleteCartsJob.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Hangfire;
+using Microsoft.Extensions.Logging;
+using VirtoCommerce.CartModule.Core;
+using VirtoCommerce.CartModule.Data.Repositories;
+using VirtoCommerce.Platform.Core.Settings;
+
+namespace VirtoCommerce.CartModule.Web.BackgroundJobs
+{
+    /// <summary>
+    /// This background job hardly removes previously softly removed carts
+    /// </summary>
+    public class DeleteObsoleteCartsJob
+    {
+        private readonly Func<ICartRepository> _repositoryFactory;
+        private readonly ISettingsManager _settingsManager;
+        private readonly ILogger _log;
+
+        public DeleteObsoleteCartsJob(Func<ICartRepository> repositoryFactory, ISettingsManager settingsManager, ILogger<DeleteObsoleteCartsJob> log)
+        {
+            _repositoryFactory = repositoryFactory;
+            _settingsManager = settingsManager;
+            _log = log;
+        }
+
+        [DisableConcurrentExecution(10)]
+        public async Task Process()
+        {
+            _log.LogTrace($"Start process DeleteObsoleteCartsJob job");
+
+            var takeCount = (int) _settingsManager.GetValue(ModuleConstants.Settings.General.PortionDeleteObsoleteCarts.Name, ModuleConstants.Settings.General.PortionDeleteObsoleteCarts.DefaultValue);
+
+            using (var repository = _repositoryFactory())
+            {
+                string[] cartIds;
+                do
+                {
+                    cartIds = repository.ShoppingCarts.Where(x => x.IsDeleted).Select(x => x.Id).Take(takeCount).ToArray();

```

Comment:
It is better to change do-while loop to the for() statement with evaluating the total softDeleted carts first, in order to avoid the case when carts will be created faster than the job will delete them, it can leads to the situation when the job is never ended.

Reply:
Agree